forest/jasmine/traj2stats.py

```diff
@@ -140,175 +140,712 @@ def get_nearby_locations(traj: np.ndarray) -> Tuple[dict, dict, dict]:
-            flight_d_vec = mov_vec[temp[:,0]==1]
-            pause_d_vec = mov_vec[temp[:,0]==2]
-            flight_t_vec = (temp[:,6]-temp[:,3])[temp[:,0]==1]
-            pause_t_vec = (temp[:,6]-temp[:,3])[temp[:,0]==2]
-            total_pause_time =  sum(pause_t_vec)
-            total_flight_time =  sum(flight_t_vec)
-            dist_traveled = sum(mov_vec)
-            if len(flight_d_vec)>0:
-                av_f_len = np.mean(flight_d_vec)
-                sd_f_len = np.std(flight_d_vec)
-                av_f_dur = np.mean(flight_t_vec)
-                sd_f_dur = np.std(flight_t_vec)
+        split_day_night = False
+
+    ids: Dict[str, List[int]] = {}
+    locations: Dict[int, List[List[float]]] = {}
+    tags: Dict[int, Dict[str, str]] = {}
+    if places_of_interest is not None or save_log:
+        ids, locations, tags = get_nearby_locations(traj)
+
+    obs_traj = traj[traj[:, 7] == 1, :]
+    home_lat, home_lon = locate_home(obs_traj, tz_str)
+    summary_stats: List[List[float]] = []
+    log_tags: Dict[str, List[dict]] = {}
+    if option == "hourly":
+        # find starting and ending time
+        sys.stdout.write("Calculating the hourly summary stats...\n")
+        time_list = stamp2datetime(traj[0, 3], tz_str)
+        time_list[4:6] = [0, 0]
+        start_stamp = datetime2stamp(time_list, tz_str) + 3600
+        time_list = stamp2datetime(traj[-1, 3], tz_str)
+        time_list[4:6] = [0, 0]
+        end_stamp = datetime2stamp(time_list, tz_str)
+        # start_time, end_time are exact points
+        # (if it ends at 2019-3-8 11 o'clock, then 11 shouldn't be included)
+        window = 60 * 60
+        no_windows = (end_stamp - start_stamp) // window
+    else:
+        # find starting and ending time
+        sys.stdout.write("Calculating the daily summary stats...\n")
+        time_list = stamp2datetime(traj[0, 3], tz_str)
+        time_list[3:6] = [0, 0, 0]
+        start_stamp = datetime2stamp(time_list, tz_str)
+        time_list = stamp2datetime(traj[-1, 3], tz_str)
+        time_list[3:6] = [0, 0, 0]
+        end_stamp = datetime2stamp(time_list, tz_str) + 3600 * 24
+        # if it starts from 2019-3-8 11 o'clock,
+        # then our daily summary starts from 2019-3-9)
+        window = 60 * 60 * 24
+        no_windows = (end_stamp - start_stamp) // window
+        if split_day_night:
+            no_windows *= 2
+
+    for i in range(no_windows):
+        if split_day_night:
+            i2 = i // 2
+        else:
+            i2 = i
+        start_time = start_stamp + i2 * window
+        end_time = start_stamp + (i2 + 1) * window
+        current_time_list = stamp2datetime(start_time, tz_str)
+        year, month, day, hour = current_time_list[:4]
+        # take a subset, the starting point of the last traj <end_time
+        # and the ending point of the first traj >start_time
+        index_rows = (traj[:, 3] < end_time) * (traj[:, 6] > start_time)
+
+        start_time2 = 0
+        end_time2 = 0
```

Comment:
I think it would make the code more readable if these two definitions are moved to where `start_time` and `end_time` are defined.